waku/v2/protocol/waku_store/waku_store.nim

```diff
@@ -459,24 +459,58 @@ proc query*(w: WakuStore, query: HistoryQuery, handler: QueryHandlerFunc) {.asyn
   waku_store_messages.set(response.value.response.messages.len.int64, labelValues = ["retrieved"])
   handler(response.value.response)
 
+proc queryFrom*(w: WakuStore, query: HistoryQuery, handler: QueryHandlerFunc, peer: PeerInfo): Future[bool] {.async.} =
+  # sends the query to the given peer
+  # returns true if no error occurs, otherwise false
+  let connOpt = await w.peerManager.dialPeer(peer, WakuStoreCodec)
+  echo "here"
+
+  if connOpt.isNone():
+    error "failed to connect to remote peer"
+    waku_store_errors.inc(labelValues = [dialFailure])
+    return false
+
+  await connOpt.get().writeLP(HistoryRPC(requestId: generateRequestId(w.rng),
+      query: query).encode().buffer)
+
+  var message = await connOpt.get().readLp(64*1024)
+  let response = HistoryRPC.init(message)
+
+  if response.isErr:
+    error "failed to decode response"
+    waku_store_errors.inc(labelValues = [decodeRpcFailure])
+    return false
+
+  waku_store_messages.set(response.value.response.messages.len.int64, labelValues = ["retrieved"])
+  handler(response.value.response)
+  return true
   
+
+proc queryLoop(w: WakuStore, query: HistoryQuery, handler: QueryHandlerFunc, candidateList: seq[PeerInfo]): Future[bool]  {.async.}= 
+  ## loops through the candidateList in order and sends the query to each until one of the query gets resolved successfully without error
+  ## returns false if all the requests fail
+  for peer in candidateList.items: 
+    let success = await w.queryFrom(query, handler, peer)
+    if success: return true
+
+  debug "failed to resolve the query"
+  return false
+
 proc findLastSeen*(list: seq[IndexedWakuMessage]): float = 
   var lastSeenTime = float64(0)
   for iwmsg in list.items : 
     if iwmsg.msg.timestamp>lastSeenTime: 
       lastSeenTime = iwmsg.msg.timestamp 
   return lastSeenTime
 
-proc resume*(ws: WakuStore) {.async, gcsafe.} =
+proc resume*(ws: WakuStore, peerList: Option[seq[PeerInfo]]) {.async.} =
```

Comment:
Return type maybe? success or not, maybe with number of messages stored as some form of feedback?

Reply:
Done!